### PR TITLE
 add plan cache 20220923

### PR DESCRIPTION
**Pull request by boyjoy1127**
Ref #19937.

Changes proposed in this pull request:
  -  add optimized plan cache
---

Before committing this PR, I'm sure that I have checked the following options:
- [X] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [X] I have self-reviewed the commit code.
- [X] I have passed maven check: `mvn clean install -B -T2C -DskipTests -Dmaven.javadoc.skip=true -e`.
- [X] I have made corresponding changes to the documentation.
- [X] I have added corresponding unit tests for my changes.

### CODE REVIEW

**Comment by zhaojinchao95** [comment]
@boyjoy1127 Hi, can your merge master branch and push again?

**Comment by boyjoy1127** [comment]
> 

OK

**Comment by boyjoy1127** [comment]
> 

I have fixed this.

File: kernel/sql-federation/executor/advanced/src/main/java/org/apache/shardingsphere/sqlfederation/advanced/AdvancedSQLFederationExecutor.java [hunk 1]
```diff
@@ -136,17 +141,23 @@ private AbstractSchema createSQLFederationSchema(final DriverExecutionPrepareEng
     
     @SuppressWarnings("unchecked")
     private ResultSet execute(final SelectStatementContext selectStatementContext, final ShardingSphereSchema schema, final AbstractSchema sqlFederationSchema, final Map<String, Object> parameters) {
+        initializePlanManagement(sqlFederationSchema);
```
Comment (strongduanmu) [inline kernel/sql-federation/executor/advanced/src/main/java/org/apache/shardingsphere/sqlfederation/advanced/AdvancedSQLFederationExecutor.java:144]: Can we skip sql node convert with plan cache?

File: kernel/sql-federation/executor/advanced/src/main/java/org/apache/shardingsphere/sqlfederation/advanced/AdvancedSQLFederationExecutor.java [hunk 2]
```diff
@@ -136,17 +141,23 @@ private AbstractSchema createSQLFederationSchema(final DriverExecutionPrepareEng
...
         SqlToRelConverter converter = SQLFederationPlannerUtil.createSqlToRelConverter(catalogReader, validator,
                 SQLFederationPlannerUtil.createRelOptCluster(JAVA_TYPE_FACTORY), optimizerContext.getSqlParserRule(), parserContext.getDatabaseType(), true);
-        SQLOptimizeContext optimizeContext =
-                new SQLOptimizeEngine(converter, SQLFederationPlannerUtil.createHepPlanner()).optimize(selectStatementContext.getSqlStatement());
-        Bindable<Object> executablePlan = EnumerableInterpretable.toBindable(Collections.emptyMap(), null, (EnumerableRel) optimizeContext.getBestPlan(), EnumerableRel.Prefer.ARRAY);
-        Enumerator<Object> enumerator = executablePlan.bind(new SQLFederationDataContext(validator, converter, parameters)).enumerator();
-        return new SQLFederationResultSet(enumerator, schema, sqlFederationSchema, selectStatementContext, optimizeContext.getValidatedNodeType());
+        SQLOptimizeEngine optimizer = new SQLOptimizeEngine(converter, SQLFederationPlannerUtil.createHepPlanner());
+        planManagement = new OptimizedPlanManagement(optimizer, validator, converter);
```
Comment (strongduanmu) [inline kernel/sql-federation/executor/advanced/src/main/java/org/apache/shardingsphere/sqlfederation/advanced/AdvancedSQLFederationExecutor.java:160]: Can we design OptimizedPlanManagement like SQLParserEngine? We can decide whether to use cache according to different scenarios.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: kernel/sql-federation/executor/advanced/src/main/java/org/apache/shardingsphere/sqlfederation/advanced/AdvancedSQLFederationExecutor.java (hunk 2 above)
Comment (boyjoy1127) [reply kernel/sql-federation/executor/advanced/src/main/java/org/apache/shardingsphere/sqlfederation/advanced/AdvancedSQLFederationExecutor.java:160]: May be we could enhance this later.

**Comment by TeslaCN** [comment]
Hi @boyjoy1127 @strongduanmu 
Is there any progress? Could we finish this in 5.3.0?

**Comment by strongduanmu** [comment]
This PR needs to be redesigned and agreed upon, so I'll close it first.

**Pull request closed without merging**